File: Week 8/2. Analyze Code Submission of a Programming Contest/programming_contest_analyze.py
import sys
import bisect

def time_to_seconds(time_point):
    hour, minute, second = map(int, time_point.split(':'))
    return hour * 3600 + minute * 60 + second

if __name__ == "__main__":
    total_submissions = 0
    error_submissions = 0
    error_submissions_user = {}
    total_point_of_user = {}
    submission_time = []
    
    for line in sys.stdin:
        line = line.strip()
        if line == "#":
            break
        
        user_id, problem_id, time_point, status, point = line.split(" ")
        point = int(point)
        
        total_submissions += 1
        
        if status == "ERR":
            error_submissions += 1
            error_submissions_user[user_id] = error_submissions_user.get(user_id, 0) + 1
            
        if status == "OK":
            if user_id not in total_point_of_user:
                total_point_of_user[user_id] = {}
            total_point_of_user[user_id][problem_id] = max(total_point_of_user[user_id].get(problem_id, 0), point)
            
        submission_time.append(time_to_seconds(time_point)) 
        
    submission_time.sort()
    
    for line in sys.stdin:
        line = line.strip()
        if line == "#":
            break
        
        parts = line.split()
        query = parts[0]
        
        if query == "?total_number_submissions":
            print(total_submissions)
        elif query == "?number_error_submision":
            print(error_submissions)
        elif query == "?number_error_submision_of_user":
            user_id = parts[1]
            if user_id in error_submissions_user:
                print(error_submissions_user[user_id])
            else:
                print(0)
        elif query == "?total_point_of_user":
            user_id = parts[1]
            if user_id in total_point_of_user:
                print(sum(total_point_of_user[user_id].values()))
            else:
                print(0)
        elif query == "?number_submission_period":
            start_time, end_time = parts[1], parts[2]
            
            start_time_seconds = time_to_seconds(start_time)
            end_time_seconds = time_to_seconds(end_time)
            
            lower = bisect.bisect_left(submission_time, start_time_seconds)
            upper = bisect.bisect_right(submission_time, end_time_seconds)
            print(upper - lower)

File: Week 8/2. Analyze Code Submission of a Programming Contest/programming_contest_analyze.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

int time_to_seconds(string timestamp)
{
    int hour, minute, second;
    sscanf(timestamp.c_str(), "%d:%d:%d", &hour, &minute, &second);
    return hour * 60 * 60 + minute * 60 + second;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int total_submissions = 0;
    int error_submissions = 0;
    unordered_map<string, int> error_submissions_user;
    unordered_map<string, unordered_map<string, int>> total_point_of_user;
    vector<int> submission_time;

    string line;

    while(getline(cin, line) && line != "#") {
        stringstream ss(line);
        string user_id, problem_id, time_point, status;
        int point;

        ss >> user_id >> problem_id >> time_point >> status >> point;

        total_submissions++;
        if (status == "ERR") {
            error_submissions_user[user_id]++;
            error_submissions++;
        }    

        if (status == "OK")
            total_point_of_user[user_id][problem_id] = max(total_point_of_user[user_id][problem_id], point);
        
        submission_time.push_back(time_to_seconds(time_point));
    }
    
    sort(submission_time.begin(), submission_time.end());

    while (getline(cin, line) && line != "#") {
        stringstream ss(line); 
        string query;
        ss >> query;

        if (query == "?total_number_submissions")
            cout << total_submissions << endl;
        else if (query == "?number_error_submision")
            cout << error_submissions << endl;
        else if (query == "?number_error_submision_of_user") {
            string user_id;
            ss >> user_id;
            cout << error_submissions_user[user_id] << endl;
        } else if (query == "?total_point_of_user") {
            string user_id;
            ss >> user_id;
            int total_point = 0;
            for (const auto& [problem_id, point]: total_point_of_user[user_id]) {
                total_point += point;
            }
            cout << total_point << endl;
        } else if (query == "?number_submission_period") {
            string start_time, end_time;
            ss >> start_time >> end_time;

            int start_time_seconds = time_to_seconds(start_time);
            int end_time_seconds = time_to_seconds(end_time);

            auto lower = lower_bound(submission_time.begin(), submission_time.end(), start_time_seconds);
            auto upper = upper_bound(submission_time.begin(), submission_time.end(), end_time_seconds);
            cout << upper - lower << endl;
        }
    }
}